indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts

```diff
@@ -335,6 +276,161 @@ function generateAddress(pubkeyArray: Uint8Array): string {
   return toBech32('dydx', ripemd160(sha256(pubkeyArray)));
 }
 
+async function validateSignature(
+  res: express.Response,
+  action: ComplianceAction,
+  address: string,
+  timestamp: number,
+  message: string,
+  signedMessage: string,
+  pubkey: string,
+  currentStatus?: string,
+): Promise<express.Response| undefined> {
+  if (!address.startsWith(DYDX_ADDRESS_PREFIX)) {
+    return create4xxResponse(
+      res,
+      `Address ${address} is not a valid dYdX V4 address`,
+    );
+  }
+
+  const pubkeyArray: Uint8Array = new Uint8Array(Buffer.from(pubkey, 'base64'));
+  if (address !== generateAddress(pubkeyArray)) {
+    return create4xxResponse(
+      res,
+      `Address ${address} does not correspond to the pubkey provided ${pubkey}`,
+    );
+  }
+
+  // Verify the timestamp is within GEOBLOCK_REQUEST_TTL_SECONDS seconds of the current time
+  const now = DateTime.now().toSeconds();
+  if (Math.abs(now - timestamp) > GEOBLOCK_REQUEST_TTL_SECONDS) {
+    return create4xxResponse(
+      res,
+      `Timestamp is not within the valid range of ${GEOBLOCK_REQUEST_TTL_SECONDS} seconds`,
+    );
+  }
+
+  // Prepare the message for verification
+  const messageToSign: string = `${message}:${action}"${currentStatus || ''}:${timestamp}`;
+  const messageHash: Uint8Array = sha256(Buffer.from(messageToSign));
+  const signedMessageArray: Uint8Array = new Uint8Array(Buffer.from(signedMessage, 'base64'));
+  const signature: ExtendedSecp256k1Signature = ExtendedSecp256k1Signature
+    .fromFixedLength(signedMessageArray);
+
+  // Verify the signature
+  const isValidSignature: boolean = await Secp256k1.verifySignature(
+    signature,
+    messageHash,
+    pubkeyArray,
+  );
+  if (!isValidSignature) {
+    return create4xxResponse(
+      res,
+      'Signature verification failed',
+    );
+  }
+
+  return undefined;
+}
+
+function validateSignatureKeplr(
+  res:express.Response,
+  address: string,
+  message: string,
+  signedMessage: string,
+  pubkey: string,
+): express.Response| undefined {
+  const messageToSign: string = message;
+
+  const pubKeyUint = new Uint8Array(Buffer.from(pubkey, 'base64'));
+  const signedMessageUint = new Uint8Array(Buffer.from(signedMessage, 'base64'));
```

Comment:
is there no timestamp?

Reply:
Yea, we are removing the timestamp so that the signature can be cached and reused. So that we do not need to prompt the user for signature every refresh